R/revision_analysis.R

```diff
@@ -0,0 +1,209 @@
+  } else if (length(arg) > 1) {
+    cli_abort("Not currently implementing more than one column at a time. Run each separately")
+  }
+  if (is.null(abs_change_threshold)) {
+    abs_change_threshold <- .05 * epi_arch$DT %>%
+      pull(...) %>%
+      max(na.rm = TRUE)
+  }
+  # for each time_value, get
+  #   the number of revisions
+  #   the maximum change in value (both absolute and relative)
+  #   the min lag
+  #   the max lag
+  #
+  # revision_tibble
+  keys <- key_colnames(epi_arch)
+  names(epi_arch$DT)
+
+  if (drop_nas) {
+    # if we're dropping NA's, we should recompactify
+    revision_behavior <-
+      epi_arch$DT %>%
+      filter(!is.na(!!arg[[1]])) %>%
+      arrange(across(c(geo_value, time_value, all_of(keys), version))) %>% # need to sort before compactifying
+      compactify_tibble(c(keys, version))
+  } else {
+    revision_behavior <- epi_arch$DT
+  }
+  revision_behavior <- revision_behavior %>%
+    mutate(across(c(version, time_value), list(int = as.integer))) %>%
+    mutate(lag = version_int - time_value_int) %>% # nolint: object_usage_linter
+    group_by(across(all_of(keys))) %>% # group by all the keys
+    select(-time_value_int, -version_int) %>%
+    summarize(
+      n_revisions = dplyr::n() - 1,
+      min_lag = min(lag), # nolint: object_usage_linter
+      max_lag = max(lag), # nolint: object_usage_linter
+      max_change = suppressWarnings(max(!!arg[[1]], na.rm = TRUE) - min(!!arg[[1]], na.rm = TRUE)),
+      max_rel_change = (max_change) / suppressWarnings(max(!!arg[[1]], na.rm = TRUE)), # nolint: object_usage_linter
+      time_to = time_to_x_percent(lag, !!arg[[1]], percent = percent_final_value)
+    ) %>%
+    ungroup() %>%
+    mutate(
+      # TODO the units here may be a problem
+      min_lag = as.difftime(min_lag, units = "days"), # nolint: object_usage_linter
+      max_lag = as.difftime(max_lag, units = "days"), # nolint: object_usage_linter
+      time_to_pct_final = as.difftime(time_to, units = "days") # nolint: object_usage_linter
+    ) %>%
+    select(-time_to)
+  if (print_inform) {
+    total_num <- nrow(revision_behavior) # nolint: object_usage_linter
+    cli_inform("Number of revisions:")
+    cli_inform("Min lag (time to first version):")
+    difftime_summary(revision_behavior$min_lag)
+    if (!drop_nas) {
+      total_na <- nrow(epi_arch$DT %>% filter(is.na(!!arg[[1]]))) # nolint: object_usage_linter
+      cli_inform("Fraction of all versions that are `NA`:")
+      cli_li(num_percent(total_na, nrow(epi_arch$DT)))
+    }
+    total_num_unrevised <- sum(revision_behavior$n_revisions == 0) # nolint: object_usage_linter
+    cli_inform("No revisions:")
+    cli_li(num_percent(total_num_unrevised, total_num))
+    total_quickly_revised <- sum( # nolint: object_usage_linter
+      revision_behavior$max_lag <=
+        as.difftime(quick_revision, units = "days")
+    )
+    cli_inform("Quick revisions (last revision within {quick_revision}
```

Comment:
issue?: I can't get any quickly-revised epikeytimes nor a substantial amount of no-revision epikeytimes from `archive_cases_dv_subset` even when selecting to individual signals.  There seems like there might be a bug (i) in this PR, and (ii) another in the archive code, underlying data, and/or my understanding of it.

```r
archive_cases_dv_subset$DT[, c("geo_value", "time_value", "version", "percent_cli")] %>%
  as_epi_archive(compactify = TRUE) %>%
  .$DT %>%
  group_by(geo_value, time_value) %>%
  summarize(max_lag = max(version - time_value), .groups = "drop") %>%
  with(min(max_lag))
#> Time difference of 1 days
archive_cases_dv_subset$DT[, c("geo_value", "time_value", "version", "percent_cli")] %>%
  as_epi_archive(compactify = TRUE) %>%
  revision_summary()
#> Number of revisions:
#> Min lag (time to first version):
#>      min median   mean     max
#>   4 days 4 days 5 days 17 days
# [......]
#> Quick revisions (last revision within 3 days of the `time_value`):
#> • 0 out of 2,168 (0%)
# [......]
```
- I don't understand how the minimum max_lag could be 1 day, and yet the min min lag is 4 days.
- I can sort of understand 0% quick revisions if it's just epikeytimes with no revisions or non-quick revisions.
- I don't understand why the no-revision and quick-revision stats are so low for the case data.  I thought most geos at most times weren't supposed to have revisions at all.